File: android/src/main/kotlin/com/example/mango_player/platform/MethodChannelHandler.kt
package com.example.mango_player.platform

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.example.mango_player.ijkplayer.IJKPlayerManager

class MethodChannelHandler(
    private val messenger: BinaryMessenger,
    private val playerManager: IJKPlayerManager,
    private val textureRegistryHandler: TextureRegistryHandler
) : MethodChannel.MethodCallHandler {

    private var channel: MethodChannel? = null

    fun startListening() {
        channel = MethodChannel(messenger, "com.mangoplayer/player")
        channel?.setMethodCallHandler(this)
    }

    fun stopListening() {
        channel?.setMethodCallHandler(null)
        channel = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                val uri = call.argument<String>("uri")
                val type = call.argument<String>("type")
                val headers = call.argument<Map<String, String>>("headers")
                val textureId = call.argument<Int>("textureId")?.toLong()
                if (uri != null && type != null) {
                    playerManager.initialize(uri, type, headers, result, textureId)
                } else {
                    result.error("invalid_arguments", "uri or type is null", null)
                }
            }
            "play" -> {
                playerManager.play()
                result.success(mapOf("success" to true))
            }
            "pause" -> {
                playerManager.pause()
                result.success(mapOf("success" to true))
            }
            "stop" -> {
                playerManager.stop()
                result.success(mapOf("success" to true))
            }
            "seekTo" -> {
                val position = call.argument<Int>("position")
                if (position != null) {
                    playerManager.seekTo(position.toLong())
                    result.success(mapOf("success" to true, "actualPosition" to position)) // Actual position might differ
                } else {
                    result.error("invalid_arguments", "position is null", null)
                }
            }
            "setVolume" -> {
                val volume = call.argument<Double>("volume")
                if (volume != null) {
                    playerManager.setVolume(volume.toFloat())
                    result.success(mapOf("success" to true))
                } else {
                    result.error("invalid_arguments", "volume is null", null)
                }
            }
            "setPlaybackSpeed" -> {
                val speed = call.argument<Double>("speed")
                if (speed != null) {
                    playerManager.setPlaybackSpeed(speed.toFloat())
                    result.success(mapOf("success" to true))
                } else {
                    result.error("invalid_arguments", "speed is null", null)
                }
            }
            "getPosition" -> {
                val position = playerManager.getPosition()
                result.success(mapOf("success" to true, "position" to position))
            }
            "getDuration" -> {
                val duration = playerManager.getDuration()
                result.success(mapOf("success" to true, "duration" to duration))
            }
            "dispose" -> {
                playerManager.release()
                result.success(mapOf("success" to true))
            }
            else -> result.notImplemented()
        }
    }
}

File: macos/Classes/Platform/MethodChannelHandler.swift
import FlutterMacOS
import Foundation

/// Logging helper
func LogInfo(_ message: String, file: String = #file, function: String = #function, line: Int = #line) {
    let filename = (file as NSString).lastPathComponent
    let logMessage = "[\(filename):\(line)] \(function): \(message)"
    print(logMessage)
    NSLog("%@", logMessage)
    fflush(stdout)
}

class MethodChannelHandler: NSObject {
    private var channel: FlutterMethodChannel?
    private let messenger: FlutterBinaryMessenger
    private let playerManager: FFmpegPlayerManager
    private let textureRegistryHandler: TextureRegistryHandler
    
    init(messenger: FlutterBinaryMessenger, playerManager: FFmpegPlayerManager, textureRegistryHandler: TextureRegistryHandler) {
        self.messenger = messenger
        self.playerManager = playerManager
        self.textureRegistryHandler = textureRegistryHandler
    }
    
    func startListening() {
        channel = FlutterMethodChannel(name: "com.mangoplayer/player", binaryMessenger: messenger)
        channel?.setMethodCallHandler(handle)
    }
    
    func stopListening() {
        channel?.setMethodCallHandler(nil)
        channel = nil
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        LogInfo("📨 Method called: \(call.method)")

        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any],
                  let uri = args["uri"] as? String,
                  let type = args["type"] as? String else {
                LogInfo("❌ Invalid arguments for initialize")
                result(FlutterError(code: "invalid_arguments", message: "uri or type is null", details: nil))
                return
            }
            let headers = args["headers"] as? [String: String]
            let textureId = (args["textureId"] as? NSNumber)?.int64Value

            LogInfo("🎬 initialize - uri: \(uri), type: \(type), textureId: \(String(describing: textureId))")
            playerManager.initialize(uri: uri, type: type, headers: headers, result: result, textureId: textureId)
            
        case "play":
            playerManager.play()
            result(["success": true])
            
        case "pause":
            playerManager.pause()
            result(["success": true])
            
        case "stop":
            playerManager.stop()
            result(["success": true])
            
        case "seekTo":
            guard let args = call.arguments as? [String: Any],
                  let position = args["position"] as? Int else {
                result(FlutterError(code: "invalid_arguments", message: "position is null", details: nil))
                return
            }
            playerManager.seekTo(position: Int64(position))
            result(["success": true, "actualPosition": position])
            
        case "setVolume":
            guard let args = call.arguments as? [String: Any],
                  let volume = args["volume"] as? Double else {
                result(FlutterError(code: "invalid_arguments", message: "volume is null", details: nil))
                return
            }
            playerManager.setVolume(volume: Float(volume))
            result(["success": true])
            
        case "setPlaybackSpeed":
            guard let args = call.arguments as? [String: Any],
                  let speed = args["speed"] as? Double else {
                result(FlutterError(code: "invalid_arguments", message: "speed is null", details: nil))
                return
            }
            playerManager.setPlaybackSpeed(speed: Float(speed))
            result(["success": true])
            
        case "getPosition":
            let position = playerManager.getPosition()
            result(["success": true, "position": position])
            
        case "getDuration":
            let duration = playerManager.getDuration()
            result(["success": true, "duration": duration])
            
        case "dispose":
            playerManager.release()
            result(["success": true])
            
        default:
            LogInfo("⚠️ Unknown method: \(call.method)")
            result(FlutterMethodNotImplemented)
        }
    }
}
